gh-122931: Include multiarch tuple in abi3t filenames

stefanor opened the pull request:
This permits `abi3t` stable ABI extensions for multiple architectures to be co-installed into the same directory, without clashing with each other, the same way (non-stable ABI) regular extensions can.

<!-- gh-issue-number: gh-122931 -->
* Issue: gh-122931
<!-- /gh-issue-number -->

stefanor committed:
gh-122917 Allow stable API extensions to include a multiarch tuple in the filename

This permits stable ABI extensions for multiple architectures to be
co-installed into the same directory, without clashing with each other,
the same way (non-stable ABI) regular extensions can.

It is listed below the current .abi3 suffix because setuptools will
select the first suffix containing .abi3, as the target filename.
We do this to protect older Python versions predating this patch.

stefanor committed:
gh-122931 Allow stable API extensions to include a multiarch tuple in the filename

This permits stable ABI extensions for multiple architectures to be
co-installed into the same directory, without clashing with each other,
the same way (non-stable ABI) regular extensions can.

It is listed below the current .abi3 suffix because setuptools will
select the first suffix containing .abi3, as the target filename.
We do this to protect older Python versions predating this patch.

vstinner reviewed (commented):
You should document this change in https://docs.python.org/dev/whatsnew/3.14.html (Doc/whatsnew/3.14.rst). I expect that you would elaborate a bit on the usage, how to use the feature, why it's needed, etc.

stefanor commented:
> You should document this change in https://docs.python.org/dev/whatsnew/3.14.html

Done.

FWIW, in Debian we plan to backport this to 3.13 too.

stefanor commented:
> Maybe Debian variants, Ubuntu, and Ubuntu variants?

All Debian derivatives, yes. They typically don't deviate very much, when it comes to plumbing.

stefanor commented:
> This change will slow down any "import module". I don't recall if there is a cache for that or not.

What do you want to do about that? Is there a benchmark you'd like to see results for?

I see a table of 5 entries (including NULL) increasing to 6. That is one extra item to search, when:
1. Importing C extensions from multiarch filenames.
2. Importing C extensions with no tag of any kind (`foo.so`).
3. Failing to import something because there isn't a C extension with this name.

vstinner commented:
> What do you want to do about that? Is there a benchmark you'd like to see results for?

Example of command:

```
strace -o trace python3 -c pass
```

strace output:

```
newfstatat(AT_FDCWD, "/usr/lib64/python3.12/encodings/__init__.cpython-312-x86_64-linux-gnu.so", 0x7ffe20002e50, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/lib64/python3.12/encodings/__init__.abi3.so", 0x7ffe20002e50, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/lib64/python3.12/encodings/__init__.so", 0x7ffe20002e50, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/lib64/python3.12/encodings/__init__.py", {st_mode=S_IFREG|0644, st_size=5884, ...}, 0) = 0

newfstatat(AT_FDCWD, "/usr/lib64/python3.12/encodings/__init__.py", {st_mode=S_IFREG|0644, st_size=5884, ...}, 0) = 0
openat(AT_FDCWD, "/usr/lib64/python3.12/encodings/__pycache__/__init__.cpython-312.pyc", O_RDONLY|O_CLOEXEC) = 3
```

Internally, when Python imports the ``encoding`` package, it has to do 4 fstat() calls. You can see a check on the `.abi3.so` suffix.

If we add a new entry to the import suffix, it will add one fstat() syscall per import (at least, to import a package). We should consider the impact on performance if we decide to add this feature.

freakboy3742 commented:
> I'm only aware of Debian who uses "multiarch". Do other operating systems also use it? Maybe Debian variants, Ubuntu, and Ubuntu variants?

FWIW: macOS, iOS and Android all use `multiarch` as a *configuration value* - it's used to differentiate ARM from x86_64 simulators, and devices from simulators/emulators. 

However, on iOS and Android, there's limited need to keep binary artefacts in the same folder, as any given executable can only have a single architecture's executables. In addition, in the case of iOS, the binaries need to be migrated to the Frameworks folder and named *as* Frameworks, so any "side-by-side" benefits would be lost. Any "other platform" executables need to be stripped out as part of the build process, at which point there's no naming conflict.

macOS *uses* the multiarch config value, but the value is always "darwin", and the universal binary format exists to support multiple architectures in a single binary file. I guess it *might* be useful to be able to have x86_64 and ARM64 binaries side-by-side... but there's probably only a year or two left in the official supported life of x86_64, so I don't think adding this feature will ultimately be that helpful for the macOS use case.

vstinner commented:
@stefanor: Did you consider to maintain this change as a downstream-only patch in Debian?

If you would like to make it upstream, I would suggest making it optional, disabled by default, and add a configure option to enable it. That's how I added some Fedora specific changes, such as:

* https://docs.python.org/dev/using/configure.html#cmdoption-with-platlibdir
* https://docs.python.org/dev/using/configure.html#cmdoption-with-wheel-pkg-dir

stefanor commented:
Here are some benchmarks:

[Google Sheet with Data](https://docs.google.com/spreadsheets/d/18ZENKDAnvTOn85N3s1AmA-Hu_8fR3O_Ddk1Wfb4A9_k/edit)

There is no discernable performance difference in minimal interpreter startup `python -c ''`.
```python
import time
import sys
from subprocess import check_call

t1 = time.perf_counter()
for i in range(1000):
    check_call([sys.executable, "-c", ""])
t2 = time.perf_counter()
print((t2-t1) / 1000)
```

![image](https://github.com/user-attachments/assets/fbc938ae-f607-44bf-ae5d-c643ac3b3a8f)

Looking at strace, I see only a single extra syscall.

***

We can manufacture an import-intensive benchmark:
```python
import pkgutil
import time

t1 = time.perf_counter()
for module in pkgutil.walk_packages(onerror=lambda pkg: None):
    if module.name.startswith("test."):
        continue
    if module.name.endswith(".__main__"):
        continue
    if module.name in {"antigravity", "idlelib.idle", "this", "zen"}:
        continue
    try:
        __import__(module)
    except Exception:
        pass
t2 = time.perf_counter()
print((t2 - t1))
```

![image](https://github.com/user-attachments/assets/3e4e6185-41a7-45c7-adfc-e3d120c2464b)

Analysing this, I see 116 stat syscalls on `.so` filenames jumping to 232. But in a 100ms total execution time, that's negligible, and I can't see any statistically significant results.

stefanor commented:
> I'm only aware of Debian who uses "multiarch".

Also, note that this is used on all linux platforms. The default for (non-stable ABI) extensions is to include the multiarch tuple in the extension filename. Pick a random binary wheel built in manylinux, and you'll see them.

> @stefanor: Did you consider to maintain this change as a downstream-only patch in Debian?

I would be happy to do that. Although I prefer not carrying downstream-only patches long-term if possible. Look at the mess around `dist-packages` for example.

Without this MR, there's no real point in supporting the multiarch tags in the non-stable ABI extensions. Either we get the benefit from doing it everywhere, or you say you don't need to support the feature, and we rip it out everywhere. We've got half a feature at the moment. Would you prefer it to all be behind a config argument?

> If you would like to make it upstream, I would suggest making it optional, disabled by default, and add a configure option to enable it. That's how I added some Fedora specific changes, such as:

These are not entirely Fedora-specific. And that's a good argument for always trying to upstream.

> https://docs.python.org/dev/using/configure.html#cmdoption-with-platlibdir

In Debian we use those paths for our cross-compilers. I imagine there is a scenario where it's useful to have Python headers in there.

> https://docs.python.org/dev/using/configure.html#cmdoption-with-wheel-pkg-dir

We're a happy customer of this too, It meant one less patch to carry.

stefanor commented:
> There is no discernable performance difference in minimal interpreter startup python -c ''.

Maybe 0.5% slowdown with PGO. For 1 extra syscall, I'm surprised it's visible. This may still not be significant.

![image](https://github.com/user-attachments/assets/2d9a89ea-d3a6-40f5-86fe-8d0eb754f7a4)

vstinner commented:
@erlend-aasland @encukou: Would you mind to have a look at this issue? What do you think? Should it become the default behavior, or should it be a configure option?

encukou commented:
IMO, the most important thing here is to keep this in mind for `abi4` (or whatever we name the free-threading-compatible stable ABI). There it should be the *only* option.
If we get abi4 in 3.14, it's probably fine to keep this as a downstream-only patch for one more release.

stefanor commented:
> IMO, the most important thing here is to keep this in mind for abi4 (or whatever we name the free-threading-compatible stable ABI). There it should be the only option.

Yeah, that sounds sensible.

> If we get abi4 in 3.14, it's probably fine to keep this as a downstream-only patch for one more release.

Is that expected? We've been on abi3 for quite a while.

stefanor commented:
> There it should be the only option.

If we want that for the stable ABI, do we want to do the same thing for the regular ABI? It would probably affect a lot of build systems that assume they can name their output `extension.so`

encukou commented:
> Is that expected? We've been on abi3 for quite a while.

Yes, over a decade. abi3 showing its age, and since it's incompatible free-threading we'll need a new alternative soon.
abi3 is not going away, but changing it this late in the cycle might not be worth it.

> If we want that for the stable ABI, do we want to do the same thing for the regular ABI? 

I don't see much reason to deprecate and remove that, so that we don't break the build systems you mention. (Note that even abi3/abi4 extensions will work with a bare `.so` extension, since the stable ABI is a subset of the full one.)

stefanor commented:
> FWIW, in Debian we plan to backport this to 3.13 too.

I'm going to include it in our 3.13.0 upload.

stefanor commented:
:bell:

vstinner commented:
Is this change useful outside Debian and Debian-like Linux distributions?

vstinner commented:
> We can't put the non-stable-ABI ones behind a configuration option, because that would break compatibility with existing wheels.

Can't you build Python with such new `./configure` option on Debian?

hugovk commented on Doc/whatsnew/3.16.rst:
```diff
@@ -75,6 +75,14 @@ New features
 ======================
 
+* Stable ABI extensions may now include a multiarch tuple in the
+  filename, e.g. ``foo.abi3-x86-64-linux-gnu.so``.
```
Avoid Latin abbreviations: https://devguide.python.org/documentation/style-guide/#use-simple-language
```suggestion
  filename, for example, ``foo.abi3-x86-64-linux-gnu.so``.
```

stefanor committed:
Order the multiarch-tagged extensions first

stefanor committed:
Order the multiarch-tagged extensions first

FFY00 commented:
> But, if this is not a default, then it is not safe for us to ever name C extensions with a multiarch suffix, the way we do for non-stable-ABI ones, in public distributions. Personally, I'd prefer to be consistent about this. At least for abi4.

AFAIK, you are the only ones shipping such extensions, no?

stefanor commented:
> AFAIK, you are the only ones shipping such extensions, no?

At the moment, we're the only ones with this patch. So, the only ones producing such extensions.

But, adding it *and* making it default (which this branch now does), means all Linux have a multiarch tag in linux `abi3` and `abi3t` extensions, the same way they do for non-stable-ABI extensions.

FFY00 commented:
But that's something IMO we want to discourage...

If you need to share `platlib` between installations due to the Debian guidelines, fine, but that's not something we really want people doing. Separating `platlib` is the solution people should be using, if at all possible, with your approach only as a last resort.

I am okay with supporting your use-case in a best-effort capacity, but I do not think it should be the default behavior.

FFY00 commented:
That said, I don't want to block this PR, so if the rest of the maintainers agree, please go ahead.

The only thing I ask is to add tests.

stefanor commented:
> If you need to share platlib between installations due to the Debian guidelines, fine, but that's not something we really want people doing

If that's the case, PEP3149 should be overturned. We implemented PEP3149, to allow us to do this. But stable-abi was forgotten, and when people started to use them, we ran into problems and filed this MR. From my position, PEP3149 works well, and I don't see any need to discourage its use.

> I am okay with supporting your use-case in a best-effort capacity, but I do not think it should be the default behavior.

It's currently the default behaviour for non-stable ABI. Why should stable-abi be different? Surely we want to align?

stefanor committed:
Test for presence of multiarch extension suffixes

stefanor committed:
Test for presence of multiarch extension suffixes

stefanor commented:
> The only thing I ask is to add tests.

Added a basic test for the presence of the tags.

Also, updated the NEWS text for the re-ordering of extensions.

rgommers commented:
> From my position, PEP3149 works well, and I don't see any need to discourage its use.

It is a continuous source of friction and latent bugs. From a packaging tool author as well as package-with-c-exts author perspective, it's definitely not ideal. So @FFY00 is right I think to recommend that if one starts fresh (e.g., a new distro from scratch), please don't do this.

> If that's the case, PEP3149 should be overturned. We implemented PEP3149, to allow us to do this.

That doesn't follow. Your use case is supported by that PEP, and is rooted in a long history - which is okay. Doesn't mean it's the recommended or optimal choice for new designs. Non-interleaved packages are clearly more robust, and the disk space / robustness tradeoffs are quite different today than 15-20 years ago.

> It's currently the default behaviour for non-stable ABI. Why should stable-abi be different? Surely we want to align?

That I can agree with. For design consistency, it's quite reasonable to treat default/unstable and stable ABI extension names the same. Beyond just some work and having to maintain a "it's different from 3.15 onwards" code in packaging tools, I don't see a significant downside to accepting this PR.

stefanor committed:
Fix editing error in NEWS and improve clarity

stefanor committed:
Fix editing error in NEWS and improve clarity

stefanor commented:
> It is a continuous source of friction and latent bugs. From a packaging tool author as well as package-with-c-exts author perspective, it's definitely not ideal.

Noted that you wouldn't recommand anybody else implement this interface. Being closer to it, I have a much more positive view. If you look at the problems it's solving, I really don't see any other obvious way to go about it that cause less friction. It massively reduces the amount of duplication and churn we have to do in the distro. It should also be a useful tool for anybody who wants to ship something for multiple Python releases and/or architectures.

How would you like to update the NEWS messaging to represent your view here?

If this is all something that Python wants to deprecate, we should design a replacement for it. We just had a packaging summit at PyCon. If that was something we should have discussed, I wish I'd known that.

stefanor committed:
Fix up the test

stefanor committed:
Fix up the test

rgommers commented:
> How would you like to update the NEWS messaging to represent your view here?

I think that entry is fine, it's just a factual description of the technical change in this PR. I don't think it needs to touch on anything bigger-picture.

> If this is all something that Python wants to deprecate

I certainly don't speak for Python (I'm not even a core dev), but: I don't think anyone wants to push for that. It's fine to exist.

> We just had a packaging summit at PyCon. If that was something we should have discussed, I wish I'd known that.

The key thing that seems needed is integration testing for interleaved installs CPython, so that the next time any requirements get captured in time, rather than afterwards. That came up on Discourse before, can't find the link right now but it was recent so I guess you remember?

vstinner commented:
@encukou:
> @vstinner, what would it take for you to reconsider?

This change adds 2 more stat() syscalls per Python import.

Example with `./python -c 'import pip'`.

* `import encodings` does 7 stat() syscalls before and 9 syscalls after.
* `import pip` does 7 stat() syscalls before and 9 syscalls after.

Example with `./python -m pip --version` (measure using `strace`):

* Before: 4,027 syscalls
* After: 4,106 syscall (+79)

My position didn't change. IMO it's a bad idea to add syscalls on all Unix platforms whereas only Debian, Ubuntu and variants need this change. As I wrote previously, I would prefer adding a configure option and use this option on Debian/Ubuntu.

We are working hard to reduce the Python startup time (lazy import is the latest major work on that). Adding "useless" syscalls sounds counter productive to me.

I would be perfectly fine with a configure option, so Debian/Ubuntu no longer have to maintain a patch downstream.

stefanor commented:
@vstinner: There are a few other components to this.

1. We have an opportunity to implement this for `abi4t` now, and not need the extra stat. Can we do this?
2. We have an asymmetry between `abi4` and the default non-stable ABI filenames. This is an opportunity to resolve that, with the temporary addition of an extra stat.

vstinner commented:
What are `abi4` and `abi4t`?

Python only supports `abi3` and `abi3t`: https://docs.python.org/dev/c-api/stable.html#stable-application-binary-interfaces.

stefanor commented:
That is, of course, what I meant.

It was proposed earlier in this MR, that we defer the change for `abi4`, and now we have `abi3t`, which is logically equivalent.

eli-schwartz commented:
As I mentioned the historical background for adding platform triplet information in the above linked ticket, I also happened to notice there was quite some discussion about the benefits. The background: https://github.com/python/cpython/issues/67169


I think that these are useful considerations that are still just as relevant today as they were in the past.

And we are already past the point of no return in committing to an additional syscall, I think (given we are settled on looking for abi3t.so), which means there isn't even a performance problem to be had in switching from one syscall to one syscall. So, what precisely is the objection to at least handling this?

@vstinner, can you confirm you have no objection? Previously you said your objection was about adding new syscalls, but when asked specifically about abi3t you took a detour into arguing about what the final name of abi3t is, and never really replied after that. The exact same thing happened in October of 2025, when you were asked about getting this right at least for the next ABI and instead of answering the question you said "right now there are no plans for an abi4": https://github.com/python/cpython/pull/122917#issuecomment-3463482773

abi4 is here. It is called abi3t. Debating which name to use is strange arguing over semantics. Everyone agrees it is a new ABI, with a new ABI soname, and it's the one after `abi3.so`.

What name shall we look it up using? `abi3t.so` or `abi3t-powerpc64le-linux-gnu.so`? Can this PR land if @stefanor updates it to only modify abi3t?

(@stefanor, please proactively do so anyway. I often find that people are much more resistant to "will you accept it if I XYZ" than "I did XYZ, will you accept it now". It's weird but it does help, so hopefully that will be enough here.)

rgommers commented:
> What name shall we look it up using? `abi3t.so` or `abi3t-powerpc64le-linux-gnu.so`?

If this is still changed, it really should be done yesterday, and it might arguably already be too late to do the "or" rather than _"look for `abi3t-powerpc64le-linux-gnu.so` first, fall back to `abi3t.so`"_. Because build backends and build systems are already released with `abi3t` support, and doing the unconditional change to ``abi3t-xxx.so` would need new releases of Maturin, scikit-build-core and CMake at this point (extension name is hardcoded there currently I believe). 

It's unlikely any wheels were already uploaded to PyPI, but it's not impossible - those would have to be yanked or hard-deleted (with lock files, yanking may not be enough).

vstinner commented:
> What name shall we look it up using? abi3t.so or abi3t-powerpc64le-linux-gnu.so? Can this PR land if @stefanor updates it to only modify abi3t?

Python 3.15 looks for `'.abi3t.so'` suffix:

```
$ python3.15
>>> import _imp; _imp.extension_suffixes()
['.cpython-315-x86_64-linux-gnu.so', '.abi3.so', '.abi3t.so', '.so']
```

I'm not aware of a plan to look for `'.abi3t-powerpc64le-linux-gnu.so`. Well, I mean outside this issue :-)

**UPDATE**: The `'.abi3t.so'` suffix was decided in PEP 803: https://peps.python.org/pep-0803/#filename-tag.

vstinner commented:
I'm not sure that this pull request is the best place to gather feedback on supporting another filename tag for C extensions. Someone should open a dedicated topic on discuss.python.org.

> It's unlikely any wheels were already uploaded to PyPI, but it's not impossible - those would have to be yanked or hard-deleted (with lock files, yanking may not be enough).

Changing filenames impacts many things, not just `_imp.extension_suffixes()`. For example, should setuptools be updated to include the multiarch tuple in the filename? Should pip be updated to accept it as well?

How to support Python 3.14 and older which don't support it? Should the two filenames be uploaded to PyPI for each release? Such question should be answered first.

stefanor commented:
> Can this PR land if @stefanor updates it to only modify abi3t?

@eli-schwartz: Absolutely, I can do that. Commits are cheap :)

> abi4 is here. It is called abi3t. Debating which name to use is strange arguing over semantics. Everyone agrees it is a new ABI, with a new ABI soname, and it's the one after abi3.so.

Yeah, that's my feeling too. I expected that the people involved in this PR discussion would help to make this happen for abi3t, from the starting gate.

---

> If this is still changed, it really should be done yesterday

@rgommers: Agreed. But I've been up against a hard rock here.

The biggest objection to this change has been performance (of supporting both names), and that's also what we'd get from just doing this right for abi3t from the start.

---

> For example, should setuptools be updated to include the multiarch tuple in the filename? Should pip be updated to accept it as well?

@vstinner: I did the analysis of tool support in https://github.com/python/cpython/issues/122931#issuecomment-4491021154 Yes, some work needs to happen, but I'm willing to drive it.

> How to support Python 3.14 and older which don't support it? Should the two filenames be uploaded to PyPI for each release? Such question should be answered first.

I will adjust this PR, on Eli's advice, to target abi3t only. 3.14 doesn't support `abi3t` so that isn't a concern, this question becomes moot. We only need to support one filename.

rgommers commented:
> @vstinner: I did the analysis of tool support in [#122931 (comment)](https://github.com/python/cpython/issues/122931#issuecomment-4491021154) Yes, some work needs to happen, but I'm willing to drive it.

That list still seems accurate. `packaging`/`pip`/`uv` shouldn't be affected, it's build backends/systems only AFAIK.

stefanor committed:
Only target abi3t (for expediency)

stefanor commented:
Rebased onto `main` to get the benefit of #150208 for tests.

eli-schwartz commented:
> If this is still changed, it really should be done yesterday, and it might arguably already be too late to do the "or" rather than _"look for `abi3t-powerpc64le-linux-gnu.so` first, fall back to `abi3t.so`"_. Because build backends and build systems are already released with `abi3t` support, and doing the unconditional change to ``abi3t-xxx.so` would need new releases of Maturin, scikit-build-core and CMake at this point (extension name is hardcoded there currently I believe).

I'm certain that build backends will be happy to do new releases in order to align with the technical decisions of an evolving situation in unstable, unreleased cpython. This isn't a barrier.

> It's unlikely any wheels were already uploaded to PyPI, but it's not impossible - those would have to be yanked or hard-deleted (with lock files, yanking may not be enough).

If they were uploaded then they're already broken by default since the ABI won't be frozen until rc1. Given cpython refuses to limit itself out of concern that wheels with wrong ABI / crashing extensions will have already been uploaded, it would be heavily inconsistent to limit itself out of concern that wheels with undetected / wrong filename extensions might have been uploaded to PyPI.

They're both just as "naughty" / unsupported, but the breakage if someone did the naughty thing is far worse for wrong ABI, and that's already a routine event in cpython development.

vstinner commented:
> doing the unconditional change to ``abi3t-xxx.so` would need new releases of Maturin, scikit-build-core and CMake at this point (extension name is hardcoded there currently I believe).

Again, that's why I think that a discussion on discuss.python.org would be better (to coordinate). Maybe in the Packaging category.

stefanor commented:
OK: https://discuss.python.org/t/can-we-consider-an-uncoditional-change-of-abi3t-so-to-abi3t-soabi-platform-so-for-3-15-last-minute/107919

stefanor commented:
Now that this PR is just renaming `abi3t` there is a follow-on commit in #152461 that adds tags to `abi3` (the original aim of this PR)

stefanor committed:
Keep .abi3t.so, as requested by the SC

stefanor commented:
Updated to reflect the SC request for both `.abi3t.so` and `.abi3t-TAG.so`

vstinner reviewed (commented):
The current implementation doesn't work properly on FreeBSD, because the `SOABI_PLATFORM` macro **is defined** but it's defined to an empty string.

```
$ grep SOABI_PLATFORM pyconfig.h
#define SOABI_PLATFORM ""
```

In Python:

```py
>>> import importlib.machinery
>>> importlib.machinery.EXTENSION_SUFFIXES
['.cpython-316d.so', '.cpython-316.so', '.abi3.so', '.abi3t-.so', '.abi3t.so', '.so']
>>> import sysconfig
>>> SOABI_PLATFORM = sysconfig.get_config_var("SOABI_PLATFORM")
>>> SOABI_PLATFORM
''
```

Notice that `'.abi3t-.so'` is present in `EXTENSION_SUFFIXES`.

Possible fix:

```diff
diff --git a/configure.ac b/configure.ac
index 2d6672e3231..c30d98af28a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1212,7 +1212,9 @@ AS_CASE([$ac_sys_system],
   [SOABI_PLATFORM=$PLATFORM_TRIPLET]
 )
 
-AC_DEFINE_UNQUOTED([SOABI_PLATFORM], ["${SOABI_PLATFORM}"], [Platform tag, used in binary module extension filenames.])
+if test x$SOABI_PLATFORM != x; then
+    AC_DEFINE_UNQUOTED([SOABI_PLATFORM], ["${SOABI_PLATFORM}"], [Platform tag, used in binary module extension filenames.])
+fi
 
 if test x$MULTIARCH != x; then
   MULTIARCH_CPPFLAGS="-DMULTIARCH=\\\"$MULTIARCH\\\""
```

vstinner committed:
Add SOABI_PLATFORM to test.pythoninfo

Add ALT_SOABI and SOABI_PLATFORM to test.pythoninfo. Add also
EXTENSION_SUFFIXES of importlib.machinery.

vstinner commented:
I pushed a change in your branch to log ALT_SOABI, SOABI_PLATFORM and EXTENSION_SUFFIXES in test.pythoninfo.

vstinner commented:
For example, I looked at the WASI job.

Extracts of test.pythoninfo on WASI:

```
importlib.extension_suffixes: ['.cpython-316d-wasm32-wasi.so', '.cpython-316-wasm32-wasi.so', '.abi3.so', '.abi3t-wasm32-wasi.so', '.abi3t.so', '.so']

sysconfig[ABIFLAGS]: d
sysconfig[ALT_SOABI]: cpython-316-wasm32-wasi
sysconfig[SOABI]: cpython-316d-wasm32-wasi
sysconfig[SOABI_PLATFORM]: wasm32-wasi
```

@stefanor:
> Those tags won't exist on not-Linux

This changes adds `'.abi3t-wasm32-wasi.so'` to `EXTENSION_SUFFIXES` on WASI. Is it deliberate?

vstinner commented:
pythoninfo on macOS:

```
importlib.extension_suffixes: ['.cpython-316d-darwin.so', '.cpython-316-darwin.so', '.abi3.so', '.abi3t-darwin.so', '.abi3t.so', '.so']
sysconfig[SOABI_PLATFORM]: darwin
```

The PR adds `.abi3t-darwin.so` suffix.

pythoninfo on iOS:

```
importlib.extension_suffixes: ['.cpython-316-iphonesimulator.so', '.abi3.so', '.abi3t-iphonesimulator.so', '.abi3t.so', '.so']
```

The PR adds `.abi3t-iphonesimulator.so` suffix.

stefanor commented:
Thanks for checking that Victor

> Notice that '.abi3t-.so' is present in EXTENSION_SUFFIXES.

Agreed that that's not what we want.

> This changes adds '.abi3t-wasm32-wasi.so' to EXTENSION_SUFFIXES on WASI. Is it deliberate?

I don't know much about the WASI stack, but that's probably a reasonable thing to do if one is likely to have WASI builds near native builds.

I'll add your configure change, it seems reasonable.

stefanor committed:
Only define SOABI_PLATFORM if it is set

stefanor commented:
FWIW, it looks like #152461 (which was originally stacked on top of this) may merge, including these changes and more.

encukou commented:
Closing in favour of #152461 -- there's no need to merge the additions in two steps.
Thanks for the PR!

stefanor commented:
Thank you for bringing that together!
